monetdbe/_cffi/internal.py

```diff
@@ -296,3 +285,14 @@ def get_columns(self, table: str, schema: str = 'sys') -> Iterator[Tuple[str, in
             name = ffi.string(names_p[0][i]).decode()
             type_ = types_p[0][i]
             yield name, type_
+
+
+from monetdbe._cffi.branch import newer_then_jul2021
+if not newer_then_jul2021:
+    def prepare(self, query: str) -> monetdbe_statement:
+        self._switch()
+        stmt = ffi.new("monetdbe_statement **")
+        check_error(lib.monetdbe_prepare(self._monetdbe_database, str(query).encode(), stmt))
```

Comment:
ah now i get it, you overwrite it dynamically. I'm surprised mypy doesn't complain about the signature here.

Maybe it is prettier and less confusing to do something like:

```
class MonetThing:
   def _prepare_jul2021(self, *, **):
       ...

   def _prepare_default(self, *, **):
       ...

   def prepare(self, *args, **kwargs):
        if jul2021:
            self._prepare_jul2021(*args, **kwargs)
        else:
            self._prepare_default(*args, **kwargs)
```

Reply:
In my case it happens at static runtime. While in your case it happens every time prepare is called. A small performance consideration basically.

Reply:
true, but if performance is of the highest importance people probably would not use python anyway. I would not sacrifice type consistency for one reasonably high-level function call less.

Reply:
We have the CI on default and Juli2021 to make sure that everything is working correctly. I could do it more OOP/Pythonic by using base and derived classes to do the override and maintain the 'performance'. But the reason why I prefer not to is because this  is basically a temporal compact code patch and that I **want** it to be ugly, concise and readily removed which will happen once Jul2021 is old enough. This discussion is a bit subjective I admit but I prefer to keep it.

Reply:
I guess this is more a matter of style, and I gave you my pythonic style view on the matter. I actually quite dislike OO also, the only reason I use it in the API is that the DBAPI dictates it. I don't want to push you to implement things in a certain way and I trust your judgment, so if you think this is fine then lets just do it this way.